Sprint 11

dmirain opened the pull request.

Commit:
ProgressHUD и гонка

Commit:
UIBlockingProgressHUD

Commit:
fetchObject fetchProfile

Commit:
fetchPhoto

Commit:
keychain

Commit:
ProfileViewController без storyboard

Commit:
Вёрстка кодом

Commit:
Привязал данные профайла

Commit:
небольшой рефакторинг

Commit:
Поправил отступ

Commit:
Починил проблему зависающего спинера

Commit:
Увеличил таймаут для медленного инернета

Commit:
Обновление аватарки

Helvetii951 commented on ImageFeed/Presentation/Profile/ProfileUIView.swift:
```diff
@@ -77,6 +83,8 @@ final class ProfileUIView: UIView {
         return view
     }
 
+    private let photoProcessor = RoundCornerImageProcessor(cornerRadius: 35)
+
     init() {
```
`RoundCornerImageProcessor` из `Kingfisher` скругляет именно ту картинку, которую скачал. Т.е., если он скачал большую картинку(например, 2000х2000), то он скруглит у нее только 35 поинтов(такое число указано в коде проекта), что вряд ли будет заметно. `RoundCornerImageProcessor` никак не влияет на скругление UIImageView аватарки. Поэтому нужно скруглять `UIImageView`в коде контроллера профиля.

Helvetii951 commented on ImageFeed/SceneDelegate.swift:
```diff
@@ -3,13 +3,56 @@ import UIKit
+            profileViewController: profileViewController
+        )
+    }
+    private var splashViewController: SplashViewController {
+        SplashViewController(
+            window: window!,
+            authStorage: authStorage,
+            authViewController: authViewController!,
+            tabBarViewController: tabBarViewController,
+            alertPresenter: AlertPresenterImpl()
+        )
+    }
```
При такой реализации, если мы не авторизованы, [в памяти уже будут котроллеры](https://ibb.co/kH0qHS2), до которых мы еще не добрались. В нашем случае это не суперкритично, но на больших проектах может привести к ненужному забитию памяти и непредсказуемому поведению.

dmirain replied:
А как иначе строить DI, скорее тогда из init надо убирать жирную логику, например вёрстку. Иначе мне не понятно, как тестировать, когда я не могу подменить зависимости. Читая статью https://habr.com/ru/companies/tinkoff/articles/546360/ вижу, что это не мною придуманная история, точнее принесённая из мира бэкенда. И да, если бы я сам архитектурил это приложение, я бы делал иначе. Но тут есть ограничения курса.

Helvetii951 commented on ImageFeed/Gateways/ProfileGateway.swift:
```diff
@@ -0,0 +1,75 @@
+
+            switch result {
+            case let .success(photoResponse):
+                NotificationCenter.default.post(
+                    name: Self.DidChangeNotification,
+                    object: self,
+                    userInfo: ["URL": photoResponse.profileImage.medium]
+                )
```
`object: self` можно передать как `nil`. При приеме нотификации в проекте он не используется.

dmirain replied:
Тоже обратил внимание.

Commit:
Поправил
- выделил в отдельный класс загрузку картинок
- правильно скругляю картинку
- при отправке сообщения не шлю объект, так как он далее не нужен
